File: backend/schemas/analytics.py
from pydantic import BaseModel
from typing import List, Optional

class KPICard(BaseModel):
    label: str
    value: str
    change: float
    changeLabel: str
    icon: str
    trend: str

class DashboardResponse(BaseModel):
    kpis: List[KPICard]

class CategoryData(BaseModel):
    name: str
    value: int
    percentage: float
    color: str

class HourlyData(BaseModel):
    hour: str
    count: int

class TrendData(BaseModel):
    month: str
    total: int
    theft: Optional[int] = 0
    assault: Optional[int] = 0
    robbery: Optional[int] = 0

class DistrictData(BaseModel):
    district: str
    totalCrimes: int
    resolved: int
    pending: int

class AnalyticsResponse(BaseModel):
    trends: List[TrendData]
    districts: List[DistrictData]
    categories: List[CategoryData]
    hourly: List[HourlyData]

class MapMarker(BaseModel):
    id: str
    latitude: float
    longitude: float
    crimeHead: str
    crimeGroup: str
    district: str

class MapResponse(BaseModel):
    markers: List[MapMarker]

class Node(BaseModel):
    id: str
    type: str
    label: str

class Link(BaseModel):
    source: str
    target: str
    label: str
    strength: float

class GraphResponse(BaseModel):
    nodes: List[Node]
    links: List[Link]

File: frontend/src/types/analytics.ts
export interface TrendDataPoint {
  month: string
  total: number
  theft: number
  assault: number
  robbery: number
  burglary: number
  fraud: number
}

export interface DistrictStats {
  district: string
  totalCrimes: number
  resolved: number
  pending: number
  arrestRate: number
  topCrimeType: string
}

export interface CategoryData {
  name: string
  value: number
  percentage: number
  color: string
}

export interface HourlyData {
  hour: string
  count: number
}

export interface WeeklyData {
  day: string
  morning: number
  afternoon: number
  evening: number
  night: number
}

export interface MonthlyData {
  month: string
  current: number
  previous: number
}

export interface SeasonalData {
  season: string
  total: number
  violent: number
  property: number
  cyber: number
}

export interface DashboardKPI {
  label: string
  value: number | string
  change: number
  changeLabel: string
  icon: string
  trend: 'up' | 'down' | 'stable'
}
